File: gadapter/src/main/java/com/zeyad/gadapter/fastscroll/viewprovider/DefaultScrollerViewProvider.java
package com.zeyad.gadapter.fastscroll.viewprovider;

import android.graphics.drawable.InsetDrawable;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.zeyad.gadapter.R;
import com.zeyad.gadapter.fastscroll.Utils;

public class DefaultScrollerViewProvider extends ScrollerViewProvider {

    private View bubble;
    private View handle;

    @Override
    public View provideHandleView(ViewGroup container) {
        handle = new View(getContext());
        int verticalInset = getScroller().isVertical() ? 0 : getContext().getResources().getDimensionPixelSize(R.dimen.fastscroll_handle_inset);
        int horizontalInset = !getScroller().isVertical() ? 0 : getContext().getResources().getDimensionPixelSize(R.dimen.fastscroll_handle_inset);
        InsetDrawable handleBg = new InsetDrawable(ContextCompat.getDrawable(getContext(), R.drawable.fastscroll_default_handle), horizontalInset,
                verticalInset, horizontalInset, verticalInset);
        Utils.setBackground(handle, handleBg);
        int handleWidth = getContext().getResources().getDimensionPixelSize(
                getScroller().isVertical() ? R.dimen.fastscroll_handle_clickable_width : R.dimen.fastscroll_handle_height);
        int handleHeight = getContext().getResources().getDimensionPixelSize(
                getScroller().isVertical() ? R.dimen.fastscroll_handle_height : R.dimen.fastscroll_handle_clickable_width);
        ViewGroup.LayoutParams params = new ViewGroup.LayoutParams(handleWidth, handleHeight);
        handle.setLayoutParams(params);
        return handle;
    }

    @Override
    public View provideBubbleView(ViewGroup container) {
        bubble = LayoutInflater.from(getContext()).inflate(R.layout.fastscroll_default_bubble, container, false);
        return bubble;
    }

    @Override
    public TextView provideBubbleTextView() {
        return (TextView) bubble;
    }

    @Override
    public int getBubbleOffset() {
        return (int) (
                getScroller().isVertical() ? ((float) handle.getHeight() / 2f) - bubble.getHeight()
                                           : ((float) handle.getWidth() / 2f) - bubble.getWidth());
    }

    @Override
    protected ViewBehavior provideHandleBehavior() {
        return null;
    }

    @Override
    protected ViewBehavior provideBubbleBehavior() {
        return new DefaultBubbleBehavior(new VisibilityAnimationManager.Builder(bubble).withPivotX(1f).withPivotY(1f).build());
    }
}

File: gadapter/src/main/java/com/zeyad/gadapter/fastscroll/viewprovider/DefaultScrollerViewProvider.kt
package com.zeyad.gadapter.fastscroll.viewprovider

import android.graphics.drawable.InsetDrawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.zeyad.gadapter.R
import com.zeyad.gadapter.fastscroll.setBackground

class DefaultScrollerViewProvider : ScrollerViewProvider() {

    private lateinit var bubble: View
    private lateinit var handle: View

    override val bubbleOffset: Int
        get() = (if (scroller?.isVertical!!)
            handle.height.toFloat() / 2f - bubble.height
        else
            handle.width.toFloat() / 2f - bubble.width).toInt()

    override fun provideHandleView(container: ViewGroup): View {
        handle = View(context)
        val verticalInset = if (scroller?.isVertical!!) 0 else context.resources.getDimensionPixelSize(R.dimen.fastscroll__handle_inset)
        val horizontalInset = if (!scroller?.isVertical!!) 0 else context.resources.getDimensionPixelSize(R.dimen.fastscroll__handle_inset)
        val handleBg = InsetDrawable(ContextCompat
                .getDrawable(context, R.drawable.fastscroll__default_handle), horizontalInset,
                verticalInset, horizontalInset, verticalInset)
        setBackground(handle, handleBg)
        val handleWidth = context.resources.getDimensionPixelSize(
                if (scroller?.isVertical!!) R.dimen.fastscroll__handle_clickable_width else R.dimen.fastscroll__handle_height)
        val handleHeight = context.resources.getDimensionPixelSize(
                if (scroller?.isVertical!!) R.dimen.fastscroll__handle_height else R.dimen.fastscroll__handle_clickable_width)
        val params = ViewGroup.LayoutParams(handleWidth, handleHeight)
        handle.layoutParams = params
        return handle
    }

    override fun provideBubbleView(container: ViewGroup): View {
        bubble = LayoutInflater.from(context)
                .inflate(R.layout.fastscroll__default_bubble, container, false)
        return bubble
    }

    override fun provideBubbleTextView(): TextView {
        return bubble as TextView
    }

    override fun provideHandleBehavior(): ViewBehavior? {
        return null
    }

    override fun provideBubbleBehavior(): ViewBehavior? {
        return DefaultBubbleBehavior(VisibilityAnimationManager.Builder(bubble).withPivotX(1f).withPivotY(1f).build())
    }
}
